File: implementations/python/lib/attack_tables.py
"""Precomputed attack and distance lookup tables."""

from typing import Callable

BOARD_SIZE = 8

KNIGHT_DELTAS = (
    (-2, -1), (-2, 1), (-1, -2), (-1, 2),
    (1, -2), (1, 2), (2, -1), (2, 1),
)

KING_DELTAS = (
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1),           (0, 1),
    (1, -1),  (1, 0),  (1, 1),
)

RAY_DELTAS = {
    (-1, -1): "southwest",
    (0, -1): "south",
    (1, -1): "southeast",
    (-1, 0): "west",
    (1, 0): "east",
    (-1, 1): "northwest",
    (0, 1): "north",
    (1, 1): "northeast",
}


def _build_attack_table(deltas: tuple[tuple[int, int], ...]) -> list[list[list[tuple[int, int]]]]:
    table = [[[] for _ in range(BOARD_SIZE)] for _ in range(BOARD_SIZE)]
    for row in range(BOARD_SIZE):
        for col in range(BOARD_SIZE):
            attacks = []
            for drow, dcol in deltas:
                new_row = row + drow
                new_col = col + dcol
                if 0 <= new_row < BOARD_SIZE and 0 <= new_col < BOARD_SIZE:
                    attacks.append((new_row, new_col))
            table[row][col] = attacks
    return table


def _build_ray_table(drow: int, dcol: int) -> list[list[list[tuple[int, int]]]]:
    table = [[[] for _ in range(BOARD_SIZE)] for _ in range(BOARD_SIZE)]
    for row in range(BOARD_SIZE):
        for col in range(BOARD_SIZE):
            ray = []
            new_row = row + drow
            new_col = col + dcol
            while 0 <= new_row < BOARD_SIZE and 0 <= new_col < BOARD_SIZE:
                ray.append((new_row, new_col))
                new_row += drow
                new_col += dcol
            table[row][col] = ray
    return table


def _build_distance_table(metric: Callable[[int, int], int]) -> list[list[int]]:
    table = [[0 for _ in range(64)] for _ in range(64)]
    for from_index in range(64):
        from_row, from_col = divmod(from_index, BOARD_SIZE)
        for to_index in range(64):
            to_row, to_col = divmod(to_index, BOARD_SIZE)
            row_distance = abs(from_row - to_row)
            col_distance = abs(from_col - to_col)
            table[from_index][to_index] = metric(row_distance, col_distance)
    return table


KNIGHT_ATTACKS = _build_attack_table(KNIGHT_DELTAS)
KING_ATTACKS = _build_attack_table(KING_DELTAS)
RAY_TABLES = {delta: _build_ray_table(*delta) for delta in RAY_DELTAS}
CHEBYSHEV_DISTANCE = _build_distance_table(max)
MANHATTAN_DISTANCE = _build_distance_table(lambda row_distance, col_distance: row_distance + col_distance)


def square_index(row: int, col: int) -> int:
    return row * BOARD_SIZE + col


def knight_attacks(row: int, col: int) -> list[tuple[int, int]]:
    return KNIGHT_ATTACKS[row][col]


def king_attacks(row: int, col: int) -> list[tuple[int, int]]:
    return KING_ATTACKS[row][col]


def ray_attacks(row: int, col: int, drow: int, dcol: int) -> list[tuple[int, int]]:
    return RAY_TABLES[(drow, dcol)][row][col]


def chebyshev_distance(from_square: tuple[int, int], to_square: tuple[int, int]) -> int:
    return CHEBYSHEV_DISTANCE[square_index(*from_square)][square_index(*to_square)]


def manhattan_distance(from_square: tuple[int, int], to_square: tuple[int, int]) -> int:
    return MANHATTAN_DISTANCE[square_index(*from_square)][square_index(*to_square)]

File: implementations/php/lib/AttackTables.php
<?php

namespace Chess;

final class AttackTables {
    private static ?self $instance = null;

    private array $knightAttacks = [];
    private array $kingAttacks = [];
    private array $rayTables = [];
    private array $chebyshevDistance = [];
    private array $manhattanDistance = [];

    private function __construct() {
        $this->knightAttacks = $this->buildAttackTable([
            [-2, -1], [-2, 1], [-1, -2], [-1, 2],
            [1, -2], [1, 2], [2, -1], [2, 1],
        ]);
        $this->kingAttacks = $this->buildAttackTable([
            [-1, -1], [-1, 0], [-1, 1],
            [0, -1], [0, 1],
            [1, -1], [1, 0], [1, 1],
        ]);

        foreach ([
            '-1,-1' => [-1, -1],
            '0,-1' => [0, -1],
            '1,-1' => [1, -1],
            '-1,0' => [-1, 0],
            '1,0' => [1, 0],
            '-1,1' => [-1, 1],
            '0,1' => [0, 1],
            '1,1' => [1, 1],
        ] as $key => [$drow, $dcol]) {
            $this->rayTables[$key] = $this->buildRayTable($drow, $dcol);
        }

        $this->chebyshevDistance = $this->buildDistanceTable(
            static fn(int $rowDistance, int $colDistance): int => max($rowDistance, $colDistance)
        );
        $this->manhattanDistance = $this->buildDistanceTable(
            static fn(int $rowDistance, int $colDistance): int => $rowDistance + $colDistance
        );
    }

    public static function getInstance(): self {
        if (self::$instance === null) {
            self::$instance = new self();
        }
        return self::$instance;
    }

    public function knightAttacks(int $row, int $col): array {
        return $this->knightAttacks[$row][$col];
    }

    public function kingAttacks(int $row, int $col): array {
        return $this->kingAttacks[$row][$col];
    }

    public function rayAttacks(int $row, int $col, int $drow, int $dcol): array {
        return $this->rayTables["{$drow},{$dcol}"][$row][$col];
    }

    public function chebyshevDistance(array $from, array $to): int {
        return $this->chebyshevDistance[$this->squareIndex($from[0], $from[1])][$this->squareIndex($to[0], $to[1])];
    }

    public function manhattanDistance(array $from, array $to): int {
        return $this->manhattanDistance[$this->squareIndex($from[0], $from[1])][$this->squareIndex($to[0], $to[1])];
    }

    private function buildAttackTable(array $deltas): array {
        $table = [];
        for ($row = 0; $row < 8; $row++) {
            $table[$row] = [];
            for ($col = 0; $col < 8; $col++) {
                $attacks = [];
                foreach ($deltas as [$drow, $dcol]) {
                    $newRow = $row + $drow;
                    $newCol = $col + $dcol;
                    if ($this->isValidSquare($newRow, $newCol)) {
                        $attacks[] = [$newRow, $newCol];
                    }
                }
                $table[$row][$col] = $attacks;
            }
        }
        return $table;
    }

    private function buildRayTable(int $drow, int $dcol): array {
        $table = [];
        for ($row = 0; $row < 8; $row++) {
            $table[$row] = [];
            for ($col = 0; $col < 8; $col++) {
                $ray = [];
                $newRow = $row + $drow;
                $newCol = $col + $dcol;
                while ($this->isValidSquare($newRow, $newCol)) {
                    $ray[] = [$newRow, $newCol];
                    $newRow += $drow;
                    $newCol += $dcol;
                }
                $table[$row][$col] = $ray;
            }
        }
        return $table;
    }

    private function buildDistanceTable(callable $metric): array {
        $table = [];
        for ($from = 0; $from < 64; $from++) {
            [$fromRow, $fromCol] = $this->indexToSquare($from);
            $table[$from] = [];
            for ($to = 0; $to < 64; $to++) {
                [$toRow, $toCol] = $this->indexToSquare($to);
                $rowDistance = abs($fromRow - $toRow);
                $colDistance = abs($fromCol - $toCol);
                $table[$from][$to] = $metric($rowDistance, $colDistance);
            }
        }
        return $table;
    }

    private function squareIndex(int $row, int $col): int {
        return $row * 8 + $col;
    }

    private function indexToSquare(int $index): array {
        return [intdiv($index, 8), $index % 8];
    }

    private function isValidSquare(int $row, int $col): bool {
        return $row >= 0 && $row < 8 && $col >= 0 && $col < 8;
    }
}
